asynclocalstorage.md

```diff
@@ -0,0 +1,371 @@
+  // be passed but should be ignored by implementations of this
+  // subset.
+  // On construction, the AsyncResource captures a reference to
+  // the current async context frame.
+  constructor(type: string, options? : InitOptions);
+  
+  // Synchronously calls fn in the scope of the frame captured
+  // when the AsyncResource instance was created.
+  runInAsyncScope(fn) : any;
+  
+  // Returns a new function that wraps the given fn such that when
+  // the returned function is called, fn is called in the scope of
+  // the frame captured when the AsyncResource instance was created.
+  bind(fn, thisArg?: any) : Function;
+  
+  // Creates a new AsyncResource and returns the `resource.bind(fn)`
+  // The type argument here is from the Node.js implementation and
+  // is passed as the `type` argument to the `AsyncResource`
+  // constructor. Implementations of this subset must be capable
+  // of accepting the argument but should ignore it.
+  static bind(fn, type?: string, thisArg?: any) : Function;
+}
+```
+
+## Importing `AsyncLocalStorage` and `AsyncResource`
+
+The classes should be importable/requirable using, at least the `node:async_hooks`
+specifier and may be accessed using the bare `async_hooks` specifier. For
+example:
+
+```
+import { AsyncLocalStorage, AsyncResource } from 'node:async_hooks'
+
+// or
+
+const { AsyncLocalStorage, AsyncResource } = require('node:async_hooks');
+```
+
+Neither the `AsyncLocalStorage` and `AsyncResource` constructors should
+be accessible via the global scope (`globalThis`).
+
+## Capturing the async context
+
+In general terms, an "async resource" is any task that is expected
+to capture the current async context frame and run within its scope
+at some point in the future. There are some async resources built in
+to JavaScript (e.g. promises, thenables, and microtasks), some defined
+by Web Platform APIs (e.g. timers, background tasks, unhandled rejections),
+some defined by the runtime, and others defined by the application. The
+common characterist of all async resources is that they are expected to
+capture the current async context frame and propagate it at various
+specific times.
+
+### Promises
+
+A promise itself is *not* an async resource. A promise continuation task,
+however, *is*. In other words, when we use `then()` to attach a continuation
+task to a promise, the current async context is captured and associated with
+that continuation task.
+
+(Note that this is a variance from Node.js' current implementation which treats
+the promise *itself* as the async resource, capturing the context at the moment
+the promise is created rather than when the continuation task is created.)
+
+#### Unhandled rejections and the deferred promise pattern
+
+The `unhandledrejection` and `rejectionhandled` events should, by default,
+propagate the context that was current when the promise was rejected and *not*
+when the promise was created.
```

Comment:
If we're straight-up disagreeing with current Node.js semantics, we should say so explicitly. Also, this is an area where we might iterate during the development of AsyncContext. I think we should mark the uncertainty here and link to related discussion issues. We could say, it must do one of those two things, and it should do [current hypothesis].